// @fivethree/core/esm2015/lib/buttons/buttons.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { FivPopover } from './../popover/popover.component';
import { Platform, PopoverController, ActionSheetController } from '@ionic/angular';
import { Component, Input, ContentChildren, QueryList } from '@angular/core';
import { FivButton } from '../button/button.component';
export class FivButtons {
    /**
     * @param {?} platform
     * @param {?} popoverController
     * @param {?} actionSheetController
     */
    constructor(platform, popoverController, actionSheetController) {
        this.platform = platform;
        this.popoverController = popoverController;
        this.actionSheetController = actionSheetController;
        this.slot = 'end';
        this.icon = 'more';
        this.count = 0;
        this.forceActionsheet = false;
        this.forcePopover = false;
        this.iconsOnly = true;
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @return {?}
     */
    ngAfterContentInit() { }
    /**
     * @param {?} event
     * @return {?}
     */
    onMenuClicked(event) {
        if (this.forceActionsheet) {
            this.presentActionSheet(event);
            return;
        }
        if (this.forcePopover) {
            this.presentPopover(event);
            return;
        }
        if (this.platform.is('mobile')) {
            this.presentActionSheet(event);
        }
        else {
            this.presentPopover(event);
        }
    }
    /**
     * @param {?} ev
     * @return {?}
     */
    presentPopover(ev) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            /** @type {?} */
            const popover = yield this.popoverController.create({
                component: FivPopover,
                event: ev,
                translucent: false,
                componentProps: {
                    buttons: this.buttons
                        .toArray()
                        .slice(this.count)
                        .filter((/**
                     * @param {?} button
                     * @return {?}
                     */
                    button => !button.disabled))
                }
            });
            yield popover.present();
            /** @type {?} */
            const result = yield popover.onDidDismiss();
            if (result &&
                result.data &&
                (result.data.index || result.data.index === 0)) {
                this.onButtonClicked(result.data.index + this.count);
            }
        });
    }
    /**
     * @param {?} ev
     * @return {?}
     */
    presentActionSheet(ev) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            /** @type {?} */
            const buttons = [];
            this.buttons
                .toArray()
                .slice(this.count)
                .forEach((/**
             * @param {?} button
             * @param {?} index
             * @return {?}
             */
            (button, index) => {
                if (!button.disabled) {
                    buttons.push({
                        text: button.text,
                        icon: button.icon,
                        handler: (/**
                         * @return {?}
                         */
                        () => {
                            this.onButtonClicked(this.count + index);
                        })
                    });
                }
            }));
            /** @type {?} */
            const actionSheet = yield this.actionSheetController.create({
                header: this.header,
                subHeader: this.subHeader,
                buttons: buttons
            });
            yield actionSheet.present();
        });
    }
    /**
     * @param {?} i
     * @return {?}
     */
    onButtonClicked(i) {
        this.buttons.toArray()[i].click.emit();
    }
}
FivButtons.decorators = [
    { type: Component, args: [{
                selector: 'fiv-buttons',
                template: "<ion-buttons *ngIf=\"buttons\" [slot]=\"slot\">\n\n  <fiv-button [color]=\"color\" *ngIf=\"(slot === 'start' || slot ==='primary') && count < buttons.length && text\"\n    (click)=\"onMenuClicked($event)\" [text]=\"text\"></fiv-button>\n\n  <fiv-button [color]=\"color\" *ngIf=\"(slot === 'start' || slot ==='primary') && count < buttons.length && !text\"\n    (click)=\"onMenuClicked($event)\" [icon]=\"'md-more'\"></fiv-button>\n  <ng-container *ngIf=\"iconsOnly\">\n    <fiv-button [color]=\"button.color\" (click)=\"onButtonClicked(i)\" [icon]=\"button.icon\"\n      *ngFor=\"let button of buttons.toArray() |\u00A0slice:0:count; let i = index;\"></fiv-button>\n  </ng-container>\n  <ng-container *ngIf=\"!iconsOnly\">\n    <fiv-button [color]=\"button.color\" (click)=\"onButtonClicked(i)\" [icon]=\"button.icon\" [text]=\"button.text\"\n      *ngFor=\"let button of buttons.toArray() |\u00A0slice:0:count; let i = index;\"></fiv-button>\n  </ng-container>\n  <fiv-button [color]=\"color\" *ngIf=\"(slot === 'end' || slot ==='secondary') && count < buttons.length && !text\"\n    (click)=\"onMenuClicked($event)\" [icon]=\"icon\"></fiv-button>\n  <fiv-button [color]=\"color\" *ngIf=\"(slot === 'end' || slot ==='secondary') && count < buttons.length && text\"\n    (click)=\"onMenuClicked($event)\" [text]=\"text\"></fiv-button>\n</ion-buttons>",
                styles: [":host{display:block}"]
            }] }
];
/** @nocollapse */
FivButtons.ctorParameters = () => [
    { type: Platform },
    { type: PopoverController },
    { type: ActionSheetController }
];
FivButtons.propDecorators = {
    slot: [{ type: Input }],
    subHeader: [{ type: Input }],
    header: [{ type: Input }],
    icon: [{ type: Input }],
    text: [{ type: Input }],
    color: [{ type: Input }],
    count: [{ type: Input }],
    forceActionsheet: [{ type: Input }],
    forcePopover: [{ type: Input }],
    iconsOnly: [{ type: Input }],
    buttons: [{ type: ContentChildren, args: [FivButton,] }]
};
if (false) {
    /** @type {?} */
    FivButtons.prototype.slot;
    /** @type {?} */
    FivButtons.prototype.subHeader;
    /** @type {?} */
    FivButtons.prototype.header;
    /** @type {?} */
    FivButtons.prototype.icon;
    /** @type {?} */
    FivButtons.prototype.text;
    /** @type {?} */
    FivButtons.prototype.color;
    /** @type {?} */
    FivButtons.prototype.count;
    /** @type {?} */
    FivButtons.prototype.forceActionsheet;
    /** @type {?} */
    FivButtons.prototype.forcePopover;
    /** @type {?} */
    FivButtons.prototype.iconsOnly;
    /** @type {?} */
    FivButtons.prototype.buttons;
    /**
     * @type {?}
     * @private
     */
    FivButtons.prototype.platform;
    /**
     * @type {?}
     * @private
     */
    FivButtons.prototype.popoverController;
    /** @type {?} */
    FivButtons.prototype.actionSheetController;
}
