// @fivethree/core/esm5/lib/popover/popover.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component } from '@angular/core';
import { NavParams, PopoverController } from '@ionic/angular';
var FivPopover = /** @class */ (function () {
    function FivPopover(navParams, popoverController) {
        this.navParams = navParams;
        this.popoverController = popoverController;
        this.buttons = [];
        this.buttons = navParams.get('buttons');
    }
    /**
     * @return {?}
     */
    FivPopover.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} i
     * @return {?}
     */
    FivPopover.prototype.onItemClicked = /**
     * @param {?} i
     * @return {?}
     */
    function (i) {
        this.popoverController.dismiss({ index: i });
    };
    FivPopover.decorators = [
        { type: Component, args: [{
                    selector: 'fiv-popover',
                    template: "<ion-content>\n  <ion-item (click)=\"onItemClicked(i)\" class=\"popover-item\" *ngFor=\"let button of buttons; let i = index\">\n    <ion-icon slot=\"start\" [name]=\"button.icon\"></ion-icon>\n    <ion-label>{{button.text}}</ion-label>\n  </ion-item>\n</ion-content>",
                    styles: [".popover-item{--border-style:none}.popover-item:hover{--background:rgb(240, 240, 240);cursor:pointer}"]
                }] }
    ];
    /** @nocollapse */
    FivPopover.ctorParameters = function () { return [
        { type: NavParams },
        { type: PopoverController }
    ]; };
    return FivPopover;
}());
export { FivPopover };
if (false) {
    /** @type {?} */
    FivPopover.prototype.buttons;
    /** @type {?} */
    FivPopover.prototype.navParams;
    /** @type {?} */
    FivPopover.prototype.popoverController;
}
