// @fivethree/core/esm2015/lib/router-item/router-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FivIcon } from './../icon/icon.component';
import { FivExpandable } from './../expandable/expandable.component';
import { Component, Input, HostBinding, HostListener, Optional, Host, ViewChild } from '@angular/core';
import { Router, NavigationEnd } from '@angular/router';
import { filter } from 'rxjs/operators';
export class FivRouterItem {
    /**
     * @param {?} router
     * @param {?} host
     */
    constructor(router, host) {
        this.router = router;
        this.host = host;
        this.active = false;
        this.detailIcon = 'ios-arrow-forward';
        this.disabled = false;
        this.matchChildUrl = false;
        this.navigate = true;
        /**
         * The position of the active state.
         * Default value is: `"left"`
         */
        this.position = 'left';
        /**
         * The shape of the active state.
         * Possible values are: `"line"`, `"dot"` and `"rounded"`.
         * Default value is: `"line"`
         */
        this.shape = 'line';
    }
    /**
     * @return {?}
     */
    onclick() {
        if (this.navigate && this.pageUrl) {
            this.router.navigateByUrl(this.pageUrl);
        }
    }
    /**
     * @return {?}
     */
    get classes() {
        return `${this.position} ${this.shape}`;
    }
    /**
     * @return {?}
     */
    get activeClass() {
        return this.matchActiveUrl() || this.active;
    }
    /**
     * @return {?}
     */
    get disabledClass() {
        return this.disabled;
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.router.events
            .pipe(filter((/**
         * @param {?} event
         * @return {?}
         */
        event => event instanceof NavigationEnd)), filter((/**
         * @param {?} event
         * @return {?}
         */
        (event) => event.url === this.pageUrl)))
            .subscribe((/**
         * @return {?}
         */
        () => this.openExpandableHost()));
    }
    /**
     * @return {?}
     */
    openExpandableHost() {
        /** @type {?} */
        const isActive = this.matchActiveUrl() || this.active;
        if (isActive) {
            if (this.host && !this.host.isOpen) {
                this.host.open();
            }
        }
    }
    /**
     * @return {?}
     */
    getClasses() {
        return [this.position, this.shape];
    }
    /**
     * @return {?}
     */
    matchActiveUrl() {
        return this.matchChildUrl ? this.isChildUrl() : this.isCurrentUrl();
    }
    /**
     * @return {?}
     */
    isCurrentUrl() {
        return this.router.url === this.pageUrl;
    }
    /**
     * @return {?}
     */
    isChildUrl() {
        return this.router.url.startsWith(this.pageUrl);
    }
}
FivRouterItem.decorators = [
    { type: Component, args: [{
                selector: 'fiv-router-item',
                template: "<ion-menu-toggle auto-hide=\"false\">\n    <fiv-ripple>\n        <ion-item [color]=\"color\" [detail]=\"detail\" [detailIcon]=\"detailIcon\" [disabled]=\"disabled\" [lines]=\"lines\">\n            <fiv-icon #fivIcon *ngIf=\"icon\" [name]=\"icon\"></fiv-icon>\n            <ion-label>{{text}}</ion-label>\n        </ion-item>\n    </fiv-ripple>\n</ion-menu-toggle>\n<span [ngClass]=\"getClasses()\" class=\"indicator\"></span>",
                styles: [":host{display:block;position:relative;width:100%}:host(.active){--color-active:var(--ion-color-primary)}:host(.active) ion-item{--color:var(--color-active)}:host(.line){--width-line:4px;--background-line:var(--ion-color-primary)}:host(.active) .line{display:block;background:var(--background-line)}:host.active.left .line{width:var(--width-line);height:100%;top:0}:host.active.right .line{width:var(--width-line);height:100%;top:0}.indicator{position:absolute}:host.active .dot{display:block}:host.active.dot.left{padding-left:24px}.dot{width:var(--fiv-dot-width,8px);height:var(--fiv-dot-heigt,8px);background-color:var(--fiv-dot-color,var(--ion-color-primary));top:50%;border-radius:100%;-webkit-transform:translateY(-50%);transform:translateY(-50%);right:8px;display:none}.dot.left{left:8px;right:unset}.dot.right{right:8px}:host(.rounded.left){--border-radius-rounded-left:0 50px 50px 0}:host(.rounded.left) ion-item{--border-radius:var(--border-radius-rounded-left)}:host(.rounded.right){--border-radius-rounded-right:50px 0 0 50px}:host(.rounded.right) ion-item{--border-radius:var(--border-radius-rounded-right)}:host(.active.rounded){--background-rounded:var(--ion-color-primary-rgb);--opacity-rounded:0.25}:host(.active.rounded) ion-item{--background:rgba(var(--background-rounded), var(--opacity-rounded))}.left{left:0}.right{right:0}ion-item{--border-style:none}fiv-icon{--fiv-padding-end:8px}ion-item:hover{--background-hover:var(--ion-color-light);cursor:pointer;--background:var(--background-hover)}:host(.router-item-disabled){pointer-events:none}"]
            }] }
];
/** @nocollapse */
FivRouterItem.ctorParameters = () => [
    { type: Router },
    { type: FivExpandable, decorators: [{ type: Optional }, { type: Host }] }
];
FivRouterItem.propDecorators = {
    fivIcon: [{ type: ViewChild, args: ['fivIcon',] }],
    active: [{ type: Input }],
    color: [{ type: Input }],
    detail: [{ type: Input }],
    detailIcon: [{ type: Input }],
    disabled: [{ type: Input }],
    icon: [{ type: Input }],
    lines: [{ type: Input }],
    matchChildUrl: [{ type: Input }],
    navigate: [{ type: Input }],
    pageUrl: [{ type: Input }],
    position: [{ type: Input }],
    shape: [{ type: Input }],
    text: [{ type: Input }],
    onclick: [{ type: HostListener, args: ['click',] }],
    classes: [{ type: HostBinding, args: ['class',] }],
    activeClass: [{ type: HostBinding, args: ['class.active',] }],
    disabledClass: [{ type: HostBinding, args: ['class.router-item-disabled',] }]
};
if (false) {
    /** @type {?} */
    FivRouterItem.prototype.fivIcon;
    /** @type {?} */
    FivRouterItem.prototype.active;
    /** @type {?} */
    FivRouterItem.prototype.color;
    /** @type {?} */
    FivRouterItem.prototype.detail;
    /** @type {?} */
    FivRouterItem.prototype.detailIcon;
    /** @type {?} */
    FivRouterItem.prototype.disabled;
    /** @type {?} */
    FivRouterItem.prototype.icon;
    /** @type {?} */
    FivRouterItem.prototype.lines;
    /** @type {?} */
    FivRouterItem.prototype.matchChildUrl;
    /** @type {?} */
    FivRouterItem.prototype.navigate;
    /** @type {?} */
    FivRouterItem.prototype.pageUrl;
    /**
     * The position of the active state.
     * Default value is: `"left"`
     * @type {?}
     */
    FivRouterItem.prototype.position;
    /**
     * The shape of the active state.
     * Possible values are: `"line"`, `"dot"` and `"rounded"`.
     * Default value is: `"line"`
     * @type {?}
     */
    FivRouterItem.prototype.shape;
    /** @type {?} */
    FivRouterItem.prototype.text;
    /** @type {?} */
    FivRouterItem.prototype.router;
    /**
     * @type {?}
     * @private
     */
    FivRouterItem.prototype.host;
}
