// @fivethree/core/esm5/lib/app-bar/app-bar-tab-content/app-bar-tab-content.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding, Output, EventEmitter } from '@angular/core';
var FivAppBarTabContent = /** @class */ (function () {
    function FivAppBarTabContent() {
        this.active = false;
        this.titleLayout = 'hide';
        this.selected = false;
        this.badge = -1;
        this.ionTabButtonClick = new EventEmitter();
    }
    Object.defineProperty(FivAppBarTabContent.prototype, "classes", {
        get: /**
         * @return {?}
         */
        function () {
            return "label-" + this.titleLayout;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    FivAppBarTabContent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    FivAppBarTabContent.prototype.onClick = /**
     * @return {?}
     */
    function () {
        // this.nav.navigateRoot(this.href);
        this.ionTabButtonClick.emit({
            tab: this.tab,
            href: this.href,
            selected: true
        });
    };
    FivAppBarTabContent.decorators = [
        { type: Component, args: [{
                    selector: 'fiv-app-bar-tab-content',
                    template: "<fiv-ripple (fivClick)=\"onClick()\" class=\"tab\" [ngClass]=\"{'active': active}\">\n  <fiv-icon [(indicatorValue)]=\"badge\" [name]=\"icon\"></fiv-icon>\n  <ion-label>{{name}}</ion-label>\n</fiv-ripple>",
                    styles: [":host{width:50%;height:100%;display:inline-block;position:relative}:host fiv-ripple{width:100%;height:100%;display:flex;position:relative;flex-direction:column;align-items:center;justify-content:center}:host fiv-icon{display:block;order:-1;--fiv-dot-top:-6px;--fiv-dot-right:-6px}:host ion-label{order:0}:host .tab.active{--fiv-color-icon:var(--ion-color-primary)}:host .tab.active ion-label{color:var(--ion-color-primary)}:host(.label-hide) ion-label{display:none}:host(.label-active-only) fiv-ripple:not(.active) ion-label{display:none}"]
                }] }
    ];
    /** @nocollapse */
    FivAppBarTabContent.ctorParameters = function () { return []; };
    FivAppBarTabContent.propDecorators = {
        name: [{ type: Input }],
        icon: [{ type: Input }],
        tab: [{ type: Input }],
        href: [{ type: Input }],
        active: [{ type: Input }],
        titleLayout: [{ type: Input }],
        badge: [{ type: Input }],
        classes: [{ type: HostBinding, args: ['class',] }],
        ionTabButtonClick: [{ type: Output }]
    };
    return FivAppBarTabContent;
}());
export { FivAppBarTabContent };
if (false) {
    /** @type {?} */
    FivAppBarTabContent.prototype.name;
    /** @type {?} */
    FivAppBarTabContent.prototype.icon;
    /** @type {?} */
    FivAppBarTabContent.prototype.tab;
    /** @type {?} */
    FivAppBarTabContent.prototype.href;
    /** @type {?} */
    FivAppBarTabContent.prototype.active;
    /** @type {?} */
    FivAppBarTabContent.prototype.titleLayout;
    /**
     * @type {?}
     * @private
     */
    FivAppBarTabContent.prototype.selected;
    /** @type {?} */
    FivAppBarTabContent.prototype.badge;
    /** @type {?} */
    FivAppBarTabContent.prototype.ionTabButtonClick;
}
